File: backend-dotnet/src/Presentation/App.Template.WebAPI/Controllers/ExportController.cs
using AppTemplate.Application.DTOs;
using AppTemplate.Application.Features.AuditLogManagement.Queries.GetAuditLogs;
using AppTemplate.Application.Features.DepartmentManagement.Queries.GetDepartments;
using AppTemplate.Application.Features.UserManagement.Queries.GetUsers;
using AppTemplate.Application.Interfaces;
using MediatR;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace AppTemplate.WebAPI.Controllers;

/// <summary>
/// Data export endpoints for CSV, Excel, and PDF formats
/// </summary>
[Authorize]
[ApiController]
[Route("api/export")]
public class ExportController : ControllerBase
{
    private readonly ISender _mediator;
    private readonly IExportService _exportService;
    private readonly ICurrentUserService _currentUserService;

    public ExportController(ISender mediator, IExportService exportService, ICurrentUserService currentUserService)
    {
        _mediator = mediator;
        _exportService = exportService;
        _currentUserService = currentUserService;
    }

    /// <summary>
    /// Export users to specified format
    /// </summary>
    [HttpGet("users")]
    [ProducesResponseType(typeof(FileResult), StatusCodes.Status200OK)]
    public async Task<IActionResult> ExportUsers(
        [FromQuery] string format = "xlsx",
        [FromQuery] bool? isActive = null,
        [FromQuery] long? departmentId = null,
        [FromQuery] string? search = null,
        CancellationToken cancellationToken = default)
    {
        var query = new GetUsersQuery
        {
            Page = 1,
            PageSize = 10000, // Large page size for export
            IsActive = isActive,
            DepartmentId = departmentId,
            Search = search
        };
        var usersResult = await _mediator.Send(query, cancellationToken);

        var exportData = usersResult.Items.Select(u => new
        {
            u.Id,
            u.Username,
            u.Email,
            u.FirstName,
            u.LastName,
            u.FullName,
            u.Role,
            u.DepartmentName,
            IsActive = u.IsActive ? "Yes" : "No",
            CreatedAt = u.CreatedAt.ToString("yyyy-MM-dd HH:mm:ss"),
            LastLoginAt = u.LastLoginAt?.ToString("yyyy-MM-dd HH:mm:ss") ?? "-"
        }).ToList();

        var exportResult = format.ToLower() switch
        {
            "csv" => await _exportService.ExportToCsvAsync(exportData, "users", cancellationToken),
            "pdf" => await _exportService.ExportToPdfAsync(exportData, "users", "Users Report", new PdfReportOptions
            {
                Subtitle = BuildFilterDescription(isActive, departmentId, search),
                GeneratedBy = _currentUserService.UserId ?? "System"
            }, cancellationToken),
            _ => await _exportService.ExportToExcelAsync(exportData, "users", "Users", cancellationToken)
        };

        return File(exportResult.FileStream, exportResult.ContentType, exportResult.FileName);
    }

    /// <summary>
    /// Export departments to specified format
    /// </summary>
    [HttpGet("departments")]
    [ProducesResponseType(typeof(FileResult), StatusCodes.Status200OK)]
    public async Task<IActionResult> ExportDepartments(
        [FromQuery] string format = "xlsx",
        [FromQuery] bool? isActive = null,
        [FromQuery] string? search = null,
        CancellationToken cancellationToken = default)
    {
        var query = new GetDepartmentsQuery
        {
            Page = 1,
            PageSize = 10000, // Large page size for export
            IsActive = isActive,
            Search = search
        };
        var deptResult = await _mediator.Send(query, cancellationToken);

        var exportData = deptResult.Items.Select(d => new
        {
            d.Id,
            d.Code,
            d.Name,
            d.Description,
            IsActive = d.IsActive ? "Yes" : "No",
            CreatedAt = d.CreatedAt.ToString("yyyy-MM-dd HH:mm:ss")
        }).ToList();

        var exportResult = format.ToLower() switch
        {
            "csv" => await _exportService.ExportToCsvAsync(exportData, "departments", cancellationToken),
            "pdf" => await _exportService.ExportToPdfAsync(exportData, "departments", "Departments Report", new PdfReportOptions
            {
                GeneratedBy = _currentUserService.UserId ?? "System"
            }, cancellationToken),
            _ => await _exportService.ExportToExcelAsync(exportData, "departments", "Departments", cancellationToken)
        };

        return File(exportResult.FileStream, exportResult.ContentType, exportResult.FileName);
    }

    /// <summary>
    /// Export audit logs to specified format
    /// </summary>
    [HttpGet("audit-logs")]
    [ProducesResponseType(typeof(FileResult), StatusCodes.Status200OK)]
    public async Task<IActionResult> ExportAuditLogs(
        [FromQuery] string format = "xlsx",
        [FromQuery] string? entityName = null,
        [FromQuery] string? action = null,
        [FromQuery] DateTime? fromDate = null,
        [FromQuery] DateTime? toDate = null,
        [FromQuery] int page = 1,
        [FromQuery] int pageSize = 1000,
        CancellationToken cancellationToken = default)
    {
        var query = new GetAuditLogsQuery
        {
            EntityName = entityName,
            Action = action,
            FromDate = fromDate,
            ToDate = toDate,
            Page = page,
            PageSize = Math.Min(pageSize, 10000) // Limit for export
        };
        var auditResult = await _mediator.Send(query, cancellationToken);

        var exportData = auditResult.Items.Select(a => new
        {
            a.Id,
            a.EntityName,
            a.EntityId,
            a.Action,
            a.UserId,
            Timestamp = a.Timestamp.ToString("yyyy-MM-dd HH:mm:ss"),
            AffectedColumns = a.AffectedColumns ?? "-"
        }).ToList();

        var exportResult = format.ToLower() switch
        {
            "csv" => await _exportService.ExportToCsvAsync(exportData, "audit_logs", cancellationToken),
            "pdf" => await _exportService.ExportToPdfAsync(exportData, "audit_logs", "Audit Log Report", new PdfReportOptions
            {
                Subtitle = $"Entity: {entityName ?? "All"}, Action: {action ?? "All"}",
                FromDate = fromDate,
                ToDate = toDate,
                GeneratedBy = _currentUserService.UserId ?? "System"
            }, cancellationToken),
            _ => await _exportService.ExportToExcelAsync(exportData, "audit_logs", "Audit Logs", cancellationToken)
        };

        return File(exportResult.FileStream, exportResult.ContentType, exportResult.FileName);
    }

    private static string BuildFilterDescription(bool? isActive, long? departmentId, string? search)
    {
        var filters = new List<string>();

        if (isActive.HasValue)
            filters.Add($"Status: {(isActive.Value ? "Active" : "Inactive")}");

        if (departmentId.HasValue)
            filters.Add($"Department ID: {departmentId.Value}");

        if (!string.IsNullOrWhiteSpace(search))
            filters.Add($"Search: \"{search}\"");

        return filters.Count > 0 ? string.Join(" | ", filters) : "All Users";
    }
}

File: backend-spring/api/src/main/java/apptemplate/api/controllers/ExportController.java
package apptemplate.api.controllers;

import apptemplate.application.dto.audit.AuditLogDto;
import apptemplate.application.dto.department.DepartmentDto;
import apptemplate.application.dto.user.UserDto;
import apptemplate.application.ports.services.CurrentUserService;
import apptemplate.application.ports.services.ExportService;
import apptemplate.application.ports.services.ExportService.ExportResult;
import apptemplate.application.ports.services.ExportService.PdfReportOptions;
import apptemplate.application.usecases.audit.GetAuditLogsUseCase;
import apptemplate.application.usecases.department.GetDepartmentsUseCase;
import apptemplate.application.usecases.user.GetUsersUseCase;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/export")
@RequiredArgsConstructor
@Tag(name = "Export", description = "Data export endpoints for CSV, Excel, and PDF formats")
public class ExportController {

    private final ExportService exportService;
    private final GetUsersUseCase getUsersUseCase;
    private final GetDepartmentsUseCase getDepartmentsUseCase;
    private final GetAuditLogsUseCase getAuditLogsUseCase;
    private final CurrentUserService currentUserService;

    private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    @GetMapping("/users")
    @Operation(summary = "Export users", description = "Export users to CSV, Excel, or PDF format")
    public ResponseEntity<byte[]> exportUsers(
            @RequestParam(defaultValue = "xlsx") String format,
            @RequestParam(required = false) String search,
            @RequestParam(required = false) Long departmentId,
            @RequestParam(required = false) Boolean isActive
    ) {
        // Default sort
        String sortBy = "createdAt";
        String sortDir = "desc";

        var users = getUsersUseCase.execute(search, departmentId, isActive, 1, 10000, sortBy, sortDir);
        var exportData = users.getContent().stream()
                .map(this::toUserExportDto)
                .collect(Collectors.toList());

        ExportResult result = switch (format.toLowerCase()) {
            case "csv" -> exportService.exportToCsv(exportData, "users", UserExportDto.class);
            case "pdf" -> exportService.exportToPdf(exportData, "users", "Users Report",
                    PdfReportOptions.builder()
                            .subtitle(buildUserFilterDescription(search, departmentId, isActive))
                            .generatedBy(currentUserService.getCurrentUsername().orElse("-"))
                            .build(),
                    UserExportDto.class);
            default -> exportService.exportToExcel(exportData, "users", "Users", UserExportDto.class);
        };

        return buildResponse(result);
    }

    @GetMapping("/departments")
    @Operation(summary = "Export departments", description = "Export departments to CSV, Excel, or PDF format")
    public ResponseEntity<byte[]> exportDepartments(
            @RequestParam(defaultValue = "xlsx") String format,
            @RequestParam(required = false) String search,
            @RequestParam(required = false) Boolean isActive
    ) {
        // Default sort
        String sortBy = "name";
        String sortDir = "asc";

        var departments = getDepartmentsUseCase.execute(search, isActive, 1, 10000, sortBy, sortDir);
        var exportData = departments.getContent().stream()
                .map(this::toDepartmentExportDto)
                .collect(Collectors.toList());

        ExportResult result = switch (format.toLowerCase()) {
            case "csv" -> exportService.exportToCsv(exportData, "departments", DepartmentExportDto.class);
            case "pdf" -> exportService.exportToPdf(exportData, "departments", "Departments Report",
                    PdfReportOptions.builder()
                            .generatedBy(currentUserService.getCurrentUsername().orElse("-"))
                            .build(),
                    DepartmentExportDto.class);
            default -> exportService.exportToExcel(exportData, "departments", "Departments", DepartmentExportDto.class);
        };

        return buildResponse(result);
    }

    @GetMapping("/audit-logs")
    @Operation(summary = "Export audit logs", description = "Export audit logs to CSV, Excel, or PDF format")
    public ResponseEntity<byte[]> exportAuditLogs(
            @RequestParam(defaultValue = "xlsx") String format,
            @RequestParam(required = false) String search,
            @RequestParam(required = false) String entityName,
            @RequestParam(required = false) String entityId,
            @RequestParam(required = false) String action,
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime fromDate,
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime toDate,
            @RequestParam(defaultValue = "1000") int limit
    ) {
        // Default sort
        String sortBy = "timestamp";
        String sortDir = "desc";
        
        var auditLogs = getAuditLogsUseCase.execute(
            search, entityName, entityId, null, action, fromDate, toDate, 
            1, limit, sortBy, sortDir
        );
        var exportData = auditLogs.getContent().stream()
                .map(this::toAuditLogExportDto)
                .collect(Collectors.toList());

        ExportResult result = switch (format.toLowerCase()) {
            case "csv" -> exportService.exportToCsv(exportData, "audit_logs", AuditLogExportDto.class);
            case "pdf" -> exportService.exportToPdf(exportData, "audit_logs", "Audit Log Report",
                    PdfReportOptions.builder()
                            .subtitle(String.format("Entity: %s, Action: %s",
                                    entityName != null ? entityName : "All",
                                    action != null ? action : "All"))
                            .fromDate(fromDate)
                            .toDate(toDate)
                            .generatedBy(currentUserService.getCurrentUsername().orElse("-"))
                            .build(),
                    AuditLogExportDto.class);
            default -> exportService.exportToExcel(exportData, "audit_logs", "Audit Logs", AuditLogExportDto.class);
        };

        return buildResponse(result);
    }

    private ResponseEntity<byte[]> buildResponse(ExportResult result) {
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + result.fileName() + "\"")
                .contentType(MediaType.parseMediaType(result.contentType()))
                .body(result.outputStream().toByteArray());
    }

    private String buildUserFilterDescription(String search, Long departmentId, Boolean isActive) {
        StringBuilder desc = new StringBuilder();
        if (isActive != null) {
            desc.append("Status: ").append(isActive ? "Active" : "Inactive");
        }
        if (departmentId != null) {
            if (!desc.isEmpty()) desc.append(" | ");
            desc.append("Department ID: ").append(departmentId);
        }
        if (search != null && !search.isEmpty()) {
            if (!desc.isEmpty()) desc.append(" | ");
            desc.append("Search: \"").append(search).append("\"");
        }
        return !desc.isEmpty() ? desc.toString() : "All Users";
    }

    private UserExportDto toUserExportDto(UserDto user) {
        return new UserExportDto(
                user.getId(),
                user.getUsername(),
                user.getEmail(),
                user.getFirstName(),
                user.getLastName(),
                user.getFullName(),
                user.getRole(),
                user.getDepartmentName(),
                user.isActive() ? "Yes" : "No",
                user.getCreatedAt() != null ? user.getCreatedAt().format(DATE_FORMATTER) : "-",
                user.getLastLoginAt() != null ? user.getLastLoginAt().format(DATE_FORMATTER) : "-"
        );
    }

    private DepartmentExportDto toDepartmentExportDto(DepartmentDto dept) {
        return new DepartmentExportDto(
                dept.getId(),
                dept.getCode(),
                dept.getName(),
                dept.getDescription(),
                dept.isActive() ? "Yes" : "No",
                dept.getCreatedAt() != null ? dept.getCreatedAt().format(DATE_FORMATTER) : "-"
        );
    }

    private AuditLogExportDto toAuditLogExportDto(AuditLogDto log) {
        return new AuditLogExportDto(
                log.getId(),
                log.getEntityName(),
                log.getEntityId(),
                log.getAction(),
                log.getUserId() != null ? log.getUserId().toString() : "-",
                log.getTimestamp() != null ? log.getTimestamp().format(DATE_FORMATTER) : "-",
                log.getAffectedColumns() != null ? log.getAffectedColumns() : "-"
        );
    }

    // Export DTOs with flattened fields for clean export
    public record UserExportDto(
            Long id,
            String username,
            String email,
            String firstName,
            String lastName,
            String fullName,
            String role,
            String departmentName,
            String isActive,
            String createdAt,
            String lastLoginAt
    ) {}

    public record DepartmentExportDto(
            Long id,
            String code,
            String name,
            String description,
            String isActive,
            String createdAt
    ) {}

    public record AuditLogExportDto(
            Long id,
            String entityName,
            String entityId,
            String action,
            String userId,
            String timestamp,
            String affectedColumns
    ) {}
}
